File: 0318 - Maximum Product of Word Lengths/solution.py
class Solution:
    def maxProduct(self, A: List[str]) -> int:
        d = {}
        for s in A:
            key = reduce(lambda res, c: res|(1<<ord(c)-ord('a')), s, 0)
            d[key] = max(d.get(key, 0), len(s))
        return max([d[v1]*d[v2] for v1 in d for v2 in d if not v1 & v2] or [0])

File: 0318 - Maximum Product of Word Lengths/solution.cpp
class Solution {
public:
    int maxProduct(vector<string>& A) {
        int res = 0;
        unordered_map<int, int> m;
        for(auto& s : A){
            int key = accumulate(s.begin(), s.end(), 0, [](int res, char c){return res|(1<<(c-'a'));});
            for(auto& [i, v] : m)
                if(!(key & i)) res = max(res, (int)s.size()*v);
            m[key] = max(m[key], (int)s.size());
        }
        return res;
    }
};